docs/proposal/20200506-single-controller-multitenancy.md

```diff
@@ -0,0 +1,869 @@
+
+``` go
+func NewClusterScope(params ClusterScopeParams) (*ClusterScope, error) {
+  ...
+  return &ClusterScope{
+    ...
+  }, nil
+}
+```
+
+#### Proposed Changes
+
+This proposal borrows heavily from the Service APIs specification.
+
+In the initial implementation, all new resources will be scoped at the cluster level, this is to enable delegation of
+AWS accounts whilst preventing privilege escalation as per [FR4](#FR4). Reasons for this are documented in the
+[alternatives](#alternatives) section.
+
+##### Cluster API Provider AWS v1alpha3 types
+
+<strong><em>Changed Resources</strong></em>
+* `AWSCluster`
+
+<strong><em>New Resources</strong></em>
+
+<em>Cluster scoped resources</em>
+
+* `AWSClusterStaticPrincipal` represents a static AWS tuple of credentials.
+* `AWSClusterRolePrincipal` represents an intent to assume an AWS role for cluster management.
+
+<em>Namespace scoped resources</em>
+
+* `AWSServiceAccountPrincipal` represents the use of a Kubernetes service account for access
+  to AWS using federated identity.
+
+<strong><em>Changes to AWSCluster</em></strong>
+
+A new field is added to the `AWSClusterSpec` to reference an principal. We intend to use `corev1.LocalObjectReference` in
+order to ensure that the only objects that can be references are either in the same namespace or are scoped to the
+entire cluster.
+
+``` go
+type  AWSClusterSpec  struct {
+  PrincipalRef corev1.LocalObjectReference `json:"principalRef,omitempty"`
+  AccountID string
+```
+
+An example usage would be:
+
+``` yaml
+---
+apiVersion: infrastructure.cluster.x-k8s.io/v1alpha3
+kind: AWSCluster
+metadata:
+  name: "test"
+  namespace: "test"
+spec:
+  region: "eu-west-1"
+  principalRef:
+    apiVersion: infrastructure.cluster.x-k8s.io/v1alpha3
+    kind: AWSClusterRolePrincipal
+    name: test-account-role
+---
+apiVersion: infrastructure.cluster.x-k8s.io/v1alpha3
+kind: AWSClusterRolePrincipal
+metadata:
+  name: "test-account-role"
+spec:
+...
+```
+
+The `PrincipalRef` field will be mutable in order to support `clusterctl move`
+scenarios where a user instantiates a cluster on their laptop and then makes
+the cluster self-managed.
+
+<strong><em>Principal CRDs</em></strong>
+
+<em>Common elements</em>
+
+All AWSCluster*Principal types will have Spec structs with an `AllowedNamespaces`
+field as follows:
+
+``` go
+type AWSClusterPrincipalSpec struct {
+  // AllowedNamespaces is a selector of namespaces that AWSClusters can
+  // use this ClusterPrincipal from. This is a standard Kubernetes LabelSelector,
+  // a label query over a set of resources. The result of matchLabels and
+  // matchExpressions are ANDed. Controllers must not support AWSClusters in
+  // namespaces outside this selector.
+  //
+  // An empty selector (default) indicates that AWSClusters can use this
+  // AWSClusterPrincipal from any namespace. This field is intentionally not a
+  // pointer because the nil behavior (no namespaces) is undesirable here.
```

Comment:
Is this what we actually want? I know this goes a little against k/k conventions, but for Cluster API we have historically disallowed empty selectors to match all.

Reply:
This is the Service API convention from https://github.com/kubernetes-sigs/service-apis/blob/master/api/v1alpha1/gatewayclass_types.go#L66